feat: improve multipart handling

Dakskihedron opened the pull request:
### Description
Resolves #13 by improving support for requests with the `multipart/form-data` content type.

### Changelog
* Updated Node version to 20.
* Add validation for file and non-file fields in the request body.
* Add file uploads using the Multer middleware for Express.js.

Dakskihedron committed:
chore: update node version

Dakskihedron committed:
feat: improve multipart support

Adds handling and validation for multipart/form-data requests, and file uploading via the Multer middleware.

karlvr commented on src/index.ts:
```diff
@@ -40,6 +40,38 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+			}
+
+			for (const op of value) {
+				if (op.requestBody?.defaultContent.mediaType.mimeType === 'multipart/form-data') return true
```
Please add a `{}` block around the `return`

karlvr commented on src/index.ts:
```diff
@@ -40,6 +40,38 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+		})
+
+		hbs.registerHelper('isSchemaSupportsMultipart', function(value: CodegenSchema): boolean {
+			return !!value.nativeType.toString().match('Api\\.(\\w+)\\.MultipartFormData(\\.\\w*)?')
```
I don't think this test will work... that `Api` piece is customisable I think. And this relies upon that `MultipartFormData` naming... so I think we'll need to come up with another way of checking this.

karlvr commented on src/index.ts:
```diff
@@ -40,6 +40,38 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+			return !!value.nativeType.toString().match('Api\\.(\\w+)\\.MultipartFormData(\\.\\w*)?')
+		})
+
+		hbs.registerHelper('isFileSchema', function(value: CodegenProperty): boolean {
```
I think this should probably be added to the Handlebars generator itself, as file is a core type.

Dakskihedron replied:
I think I could remove this instead cos it's only used in one place and it functions the same as passing `schema` to `isFile`?

karlvr commented on src/index.ts:
```diff
@@ -55,6 +87,14 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
 				{ ...rootContext, ...group, ...doc }, false, hbs)
+
+			for (const operation of operations) {
+				if (operation.requestBody?.defaultContent.mediaType.mediaType === 'multipart/form-data') {
```
Maybe we can pull the code out of `hasOperationSupportingMultipart` into a function and re-use it here. It feels like we should have one way of determining this.

karlvr commented on templates/api.hbs:
```diff
@@ -1,17 +1,25 @@
 		{{>frag/apiSecurityRequirements}}
-		function (req, res) {
+		{{#if (hasOperationSupportingMultipart .)}}
+		f.upload.fields([{{#with (getFileProperties requestBody.defaultContent.schema.properties)}}{{#each this}}{name: {{{stringLiteral @key}}}, maxCount: 1}{{#hasMore}}, {{/hasMore}}{{/each}}{{/with}}]),
```
That `maxCount`... there can be file arrays, which I think we're probably not handling here.

karlvr commented on templates/api.hbs:
```diff
@@ -1,17 +1,25 @@
-		function (req, res) {
+		{{#if (hasOperationSupportingMultipart .)}}
+		f.upload.fields([{{#with (getFileProperties requestBody.defaultContent.schema.properties)}}{{#each this}}{name: {{{stringLiteral @key}}}, maxCount: 1}{{#hasMore}}, {{/hasMore}}{{/each}}{{/with}}]),
+		f.PostLogWithId.handleUpload,
```
This line looks like a mistake.

karlvr commented on templates/api.hbs:
```diff
@@ -32,7 +40,9 @@ export default function(app: Express, impl: t.{{className name}}Api) {
 						{{else}}
-						{{#ifmatch mediaType.mimeType '/json'}}
+						{{#ifmatch mediaType.mimeType 'multipart/form-data'}}
+						return {{>frag/fromJson schema prefix='v.'}}('body', req)
```
Perhaps it would be easier to duplicate `fromJson.hbs` and make a multipart version?

karlvr commented on templates/frag/fromJson.hbs:
```diff
@@ -13,8 +13,10 @@ Output a reference to a FromJsonFunction function.
+{{prefix}}fileFromMultipart
 {{~else if (isObject)~}}
-{{prefix}}model{{className nativeType.parentType}}FromJson
+{{prefix}}model{{className nativeType.parentType}}From{{#if (isSchemaSupportsMultipart .)}}MultipartFormData{{else}}Json{{/if}}
```
I feel like this logic might not be right... if we've received a JSON request body then we want to parse it as JSON no matter whether the schema _supports_ multipart, such as in the case where one schema could be received as multipart or JSON.

My comment above about splitting `fromJson.hbs` into `fromJson.hbs` and `fromMultipart.hbs` will fix this I think.

karlvr commented on templates/frag/parseMultipartValue.hbs:
```diff
@@ -0,0 +1,11 @@
+Determine the correct value to pass from a multipart property to the FromJsonFunction.
+@param <this> CodegenProperty or CodegenSchema
+--}}
+{{#if (isFileSchema .)}}
```
I need you to explain this file to me :-D

Dakskihedron replied:
A multipart request is made up of a `body` containing all name-value pairs and `files` containing all name-file pairs. This file helps determine what value to use at each level of validation (and also prevents broken indentation).

If the value passes
- `isFileSchema`, then the value is a `Express.Multer.File[]` array containing the file object we want;
- `isSchemaSupportsMultipart`, then the value is the `files` part of the request made up of name and `Express.Multer.File[]` values;
- else the value must be the `body` part of the request.

karlvr commented on templates/validationGeneric.hbs:
```diff
@@ -1,6 +1,6 @@
 const {{constantName (concat nativeType 'Keys')}}: string[] = [{{#each properties}}{{{stringLiteral serializedName}}}{{#hasMore}}, {{/hasMore}}{{/each}}]
 
-function model{{className nativeType.parentType}}FromJsonContent(name: string, value: any, knownKeys: Record<string, boolean> = {}): {{#if discriminator}}Omit<{{nativeType.parentType}}, {{{stringLiteral discriminator.serializedName}}}>{{else}}{{nativeType}}{{/if}} {
+function model{{className nativeType.parentType}}From{{#if (isSchemaSupportsMultipart .)}}MultipartFormData{{else}}Json{{/if}}Content(name: string, value: any, knownKeys: Record<string, boolean> = {}): {{#if discriminator}}Omit<{{nativeType.parentType}}, {{{stringLiteral discriminator.serializedName}}}>{{else}}{{nativeType}}{{/if}} {
```
I feel like we need separate multipart and JSON versions for these... in relation to my comments about `fromJson.hbs`... and it might be a bit easier to reason about.

karlvr commented on templates/validationGeneric.hbs:
```diff
@@ -36,6 +58,7 @@ function model{{className nativeType.parentType}}FromJsonContent(name: string, v
 	return result
 }
 
+{{#unless (isSchemaSupportsMultipart .)}}
```
I'm not sure we need to exclude this function? The schema might still be output as JSON. Maybe it's no problem to generate this function even if it's not used?

karlvr reviewed (changes requested):
Looking really exciting and feeling pretty close to done!

Dakskihedron committed:
fix: remove hardcoded API class name

Dakskihedron committed:
feat: move multipart validation into new file

Dakskihedron committed:
refactor: remove hacks from generator code

Or at least try my best to.

Dakskihedron committed:
refactor: move duplicate code into its own func

Dakskihedron committed:
feat: support arrays of files

First step in adding support for validation of file arrays and imposing max array length.

Dakskihedron committed:
feat: add middleware to enforce minItems

Add a middleware function that enforces the minimum number of files required in a file array.

Dakskihedron committed:
feat: add support for nullable items in arrays

karlvr commented on src/index.ts:
```diff
@@ -40,6 +40,52 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
 		})
 
+		hbs.registerHelper('hasOperationSupportingMultipart', function(value: CodegenOperation | CodegenOperation[]): boolean {
+			if (isCodegenOperation(value)) {
```
Let's use the `containsMultipartOperation` function here too. It has the logic we need.

e.g.

```typescript
return containsMultipartOperation([value])
```

karlvr commented on src/index.ts:
```diff
@@ -40,6 +40,52 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+				return value.requestBody?.defaultContent.mediaType.mimeType === 'multipart/form-data'
+			}
+
+			return containsMultipartOperation(value)
```
I think this would be nice in an `else` block.

karlvr commented on src/index.ts:
```diff
@@ -40,6 +40,52 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+		})
+
+		hbs.registerHelper('isMultipartSchema', function(value: CodegenSchema): boolean {
+			return value.originalName === 'multipart/form-data'
```
I don't like relying on the name... so I guess we need to have a way to see that this is a special schema (maybe a purpose) _and_ that it was generated for a particular mime-type encoding.

Dakskihedron replied:
I've implemented a potential solution: https://github.com/karlvr/openapi-generator-plus/pull/56

karlvr commented on src/index.ts:
```diff
@@ -40,6 +40,52 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+			return value.purpose === CodegenSchemaPurpose.METADATA
+		})
+
+		hbs.registerHelper('getFileProperties', function(properties: CodegenObjectLikeSchemas[]): FileProperty[] {
```
perhaps this should just be `fileProperties`... does that match other helper naming?

Dakskihedron replied:
I might need to rethink this one entirely. The intention is to tell the user it gets all properties that are files.

Dakskihedron replied:
I feel like keeping the `get` in the helper name makes it obvious that it retrieves something.

karlvr commented on src/index.ts:
```diff
@@ -40,6 +40,52 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+					continue
+				}
+
+				if (!(property.schema.purpose === CodegenSchemaPurpose.METADATA)) {
```
This can use `!==` instead.

karlvr commented on src/index.ts:
```diff
@@ -150,4 +203,29 @@ function compareOperations(a: CodegenOperation, b: CodegenOperation): number {
+/**
+ * An interface representing file properties in a multipart/form-data request.
+ */
+type FileProperty = {
```
This should be `interface FileProperty {` instead... usually I'd put interface definitions at the top of the file.

karlvr commented on templates/api.hbs:
```diff
@@ -1,17 +1,39 @@
 		{{{stringLiteral (pathTemplate fullPath)}}},
 		{{>frag/apiSecurityRequirements}}
-		function (req, res) {
+		{{#if (hasOperationSupportingMultipart .)}}
```
Curious that we don't need to check for multipart, we only need to check for `FILE` as they're always multipart. That's curious because they're sort of separate concepts... it's not clear that `FILE` is only multi-part, but we set the type to `Express.Multer.File` in our `index.ts`... HMM

karlvr commented on templates/api.hbs:
```diff
@@ -1,17 +1,39 @@
+		(req, res, next) => {
+			f.upload.fields([{{#each this}}{ name: {{{stringLiteral name}}}, maxCount: {{#if maxCount}}{{{maxCount}}}{{else}}1{{/if}} }{{#hasMore}}, {{/hasMore}}{{/each}}])(req, res, (error) => {
+				if (error?.code === 'LIMIT_UNEXPECTED_FILE') {
+					res.status(400).send(`You have exceeded the maximum number of files allowed for the field '${error.field}'`)
```
Let's include the maximum number of files in the error message.

```typescript
res.status(400).send(`The field '${fieldName}' allows a maximum of {{#if maxCount}}{{{maxCount}}}{{else}}1{{/if}} file(s)`)
```

This also syncs up the message with the minimum. I preferred the minimum message style as "you" is a bit personal for an API message.

Dakskihedron replied:
Bit of an issue with this approach: the error returned by Multer does not contain the max number of files allowed for a field.

Dakskihedron replied:
I've devised a solution for this.

karlvr commented on templates/api.hbs:
```diff
@@ -1,17 +1,39 @@
+			f.upload.fields([{{#each this}}{ name: {{{stringLiteral name}}}, maxCount: {{#if maxCount}}{{{maxCount}}}{{else}}1{{/if}} }{{#hasMore}}, {{/hasMore}}{{/each}}])(req, res, (error) => {
+				if (error?.code === 'LIMIT_UNEXPECTED_FILE') {
+					res.status(400).send(`You have exceeded the maximum number of files allowed for the field '${error.field}'`)
+					return
```
We don't need this `return` here.

karlvr commented on templates/api.hbs:
```diff
@@ -1,17 +1,39 @@
+			})
+		},
+		{{~#each this}}
+		{{#if minCount}}v.verifyMinimumFiles({{{stringLiteral name}}}, {{{minCount}}}),{{/if}}
```
I think the `{{#if}}` should be on its own line, and the `{{/if}}` so that if it's false we don't get a blank line.

Handlebars strips lines if they _only_ contain a block command, ie. `{{#...}`

karlvr commented on templates/api.hbs:
```diff
@@ -32,7 +54,9 @@ export default function(app: Express, impl: t.{{className name}}Api) {
-						{{#ifmatch mediaType.mimeType '/json'}}
+						{{#ifmatch mediaType.mimeType 'multipart/form-data'}}
+						return {{>frag/fromMultipart schema prefix='v.'}}('body', req)
+						{{else if (ifmatch mediaType.mimeType '/json')}}
```
I'd like to use `(isContentJson .)` here instead of `(ifmatch mediaType.mimeType '/json')`

karlvr commented on templates/api.hbs:
```diff
@@ -32,7 +54,9 @@ export default function(app: Express, impl: t.{{className name}}Api) {
 						return undefined
 						{{else}}
-						{{#ifmatch mediaType.mimeType '/json'}}
+						{{#ifmatch mediaType.mimeType 'multipart/form-data'}}
```
I'd like to use `{{#if (isContentMultipart .)}` here instead of `{{#ifmatch mediaType.mimeType 'multipart/form-data'}}`

karlvr commented on templates/apiMultipartHelper.hbs:
```diff
@@ -0,0 +1,18 @@
+{{#each operations}}
+{{#if (hasOperationSupportingMultipart .)}}
+export namespace {{className name}} {
+    /* Middleware entry point */
```
Let's make this a jsdoc comment (`/**`) and perhaps it should say:

```typescript
/** Special handling for uploads to the {{{name}}} operation. */
```

karlvr commented on templates/apiMultipartHelper.hbs:
```diff
@@ -0,0 +1,18 @@
+
+{{#each operations}}
+{{#if (hasOperationSupportingMultipart .)}}
+export namespace {{className name}} {
```
Does this include the group name of the operation? Maybe we should have an outer namespace for the operation group name.

karlvr commented on templates/frag/fromMultipart.hbs:
```diff
@@ -0,0 +1,42 @@
+{{!--
+Output a reference to a FromJsonFunction function.
```
This comment needs to change...

```
Output a reference to a function to transform the multipart input value into a JavaScript value
```

We could change the comment in `fromJson.hbs` to be clearer too:

```
Output a reference to a function to transform the JSON input value into a JavaScript value
```

karlvr commented on templates/frag/parseMultipartValue.hbs:
```diff
@@ -0,0 +1,11 @@
+{{!--
+Determine the correct value to pass from a multipart property to the FromJsonFunction.
```
I think "pass" should be "parse". Maybe we need the same change in the JSON equivalent? I think this file should be named `extract` rather than `parse` actually.

```
Extract the value from the multipart input.
```

Dakskihedron replied:
There's no JSON equivalent for this.

karlvr commented on templates/validation.hbs:
```diff
@@ -303,5 +325,55 @@ export function timeToJson(name: string, value: string): string {
+	if (Array.isArray(value)) {
+		return value[0]
+	}
+	return value
```
Let's put this in an `else` block.

karlvr commented on templates/validationMultipart.hbs:
```diff
@@ -0,0 +1,130 @@
+	{{/if}}
+}
+
+export function model{{className nativeType.parentType}}ToJson(name: string, value: {{nativeType}}): ToJson<{{nativeType}}> {
```
Do we use these `toJson` functions? Are they different from the ones in the json validation?

Dakskihedron replied:
These functions are the same as those used in JSON validation. I don't believe they get used.

karlvr commented on templates/validationMultipart.hbs:
```diff
@@ -0,0 +1,130 @@
+	return result
+}
+
+function model{{className nativeType.parentType}}ToJsonContent(name: string, value: {{nativeType}}, knownKeys: Record<string, boolean> = {}): ToJson<{{#if discriminator}}Omit<{{nativeType.parentType}}, {{{stringLiteral discriminator.serializedName}}}>{{else}}{{nativeType}}{{/if}}> {
```
Do we use these `toJson` functions? Are they different from the ones in the json validation?

Dakskihedron replied:
These functions are the same as those used in JSON validation. I don't believe they get used.

Dakskihedron committed:
refactor: improve error handling for file upload

Error message has the field name and max file count when maximum is exceeded.

Dakskihedron committed:
refactor: use more descriptive names and comments

Dakskihedron committed:
feat: improve namespace organisation

Dakskihedron committed:
fix: correct helper name

Dakskihedron committed:
refactor: improve code structure and reuse

Dakskihedron committed:
refactor: simplify code

Simplifying the code also fixes broken indentations in the generator output.

Dakskihedron committed:
refactor: move toJson functions to frag

Dakskihedron committed:
feat: improve code robustness and naming

Add type checks and better comparisons. Use more appropriate names for multipart functions.

karlvr commented on src/index.ts:
```diff
@@ -40,6 +52,52 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+			return value.purpose === CodegenSchemaPurpose.METADATA
+		})
+
+		hbs.registerHelper('getFileUploadProperties', function(properties: CodegenObjectLikeSchemas[]): FileUploadProperty[] {
```
What do you think about the name `toFileUploadProperties` instead? It's transforming its input rather than retrieving something.

karlvr commented on src/index.ts:
```diff
@@ -40,6 +52,52 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+			const results: FileUploadProperty[] = []
+			for (const prop in properties) {
+				const property = properties[prop]
+				if (!isCodegenSchemaUsage(property)) {
```
Is this check necessary? The type of `property` should be `CodegenProperty` so I thought this would be redundant.

Dakskihedron replied:
I've removed the need for that first check by changing the input type to `CodegenProperty`.

karlvr commented on src/index.ts:
```diff
@@ -40,6 +52,52 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+					continue
+				}
+
+				if (isCodegenObjectSchema(property.schema) && property.schema.properties) {
```
Do we need this test? I wonder if we instead check the array schema _first_ and then this is just the `else` block... we don't need to check it's an object schema here?

Dakskihedron replied:
Checked and it's fine.

karlvr commented on templates/frag/extractMultipartValue.hbs:
```diff
@@ -0,0 +1,11 @@
+{{~else if (isMetadataSchema schema)~}}
+value.files && !Array.isArray(value.files) && value.files[{{{stringLiteral serializedName}}}]
+{{~else~}}
+value.body[{{{stringLiteral serializedName}}}]
```
When `value` is of type `Express.Request` does this still compile?

karlvr commented on templates/frag/extractMultipartValue.hbs:
```diff
@@ -0,0 +1,11 @@
+--}}
+{{#if (isFile schema)}}
+value
+{{~else if (isMetadataSchema schema)~}}
```
I think we should instead check `(isFile schema)`... there could be metadata schemas that aren't files.

karlvr commented on templates/frag/fromJson.hbs:
```diff
@@ -13,10 +13,12 @@ Output a reference to a FromJsonFunction function.
 {{~else if (isBinary)~}}
 {{prefix}}binaryFromJson
+{{~else if (isFile)~}}
+{{prefix}}fileFromMultipart
```
Thi is in the `fromJson` frag so we can't support the `FILE` type in JSON... so let's change this to `unsupportedFromJson`

karlvr commented on templates/frag/fromMultipart.hbs:
```diff
@@ -0,0 +1,38 @@
+{{~else if (isFile)~}}
+{{prefix}}fileFromMultipart
+{{~else if (isObject)~}}
+{{prefix}}model{{className nativeType.parentType}}FromMultipart
```
OOOOOOH! I think we can drop the whole "FromMultipart" and "FromJson" thing from our naming. We generate _either_ the multipart version _or_ the JSON version for a given schema. Let's not change that now but as soon as we merge this... let's try to replace "FromJson" and "FromMultipart" with "FromValue".

Dakskihedron committed:
refactor: reduce duplicate code and improve reuse

Significant refactoring to remove duplicate functions and promote better code reusage. Also more appropriate naming.

Dakskihedron committed:
docs: improve comment

Dakskihedron committed:
refactor: code cleanup and improvements

Dakskihedron committed:
feat: remove file helper function

Remove the file helper function to encourage any post-upload handling within the main API logic instead.

karlvr commented on templates/api.hbs:
```diff
@@ -1,17 +1,40 @@
 		{{{stringLiteral (pathTemplate fullPath)}}},
 		{{>frag/apiSecurityRequirements}}
-		function (req, res) {
+		{{#if (hasOperationSupportingMultipart .)}}
```
Curiously... do we even need this check? We can probably just call `toFileUploadProperties` or even rename that `fileUploadProperties`... as it returns the file upload properties from the given properties.

karlvr commented on src/index.ts:
```diff
@@ -40,6 +52,40 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+					continue
+				}
+
+				if (isCodegenArraySchema(property.schema) && property.schema.component) {
```
I think we need to check here that `property.schema.component.schema.schemaType === CodegenSchemaType.FILE`.

Dakskihedron replied:
The correct `schemaType` is nested a bit deeper than that. We need to make a comparison against `property.schema.component.schema.properties.value.schema.schemaType`. However, this has a lot of ugly type-guards that need to be implemented, otherwise it gets annoyed.

karlvr commented on src/index.ts:
```diff
@@ -40,6 +52,40 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+						minCount: property.schema.minItems,
+						maxCount: property.schema.maxItems,
+					})
+				} else {
```
And here that `property.schema.schemaType === CodegenSchemaType.FILE`

Dakskihedron replied:
Similar idea with this one except we won't need to deal with the `component.schema` part.

karlvr commented on templates/api.hbs:
```diff
@@ -32,17 +55,27 @@ export default function(app: Express, impl: t.{{className name}}Api) {
-						{{#ifmatch mediaType.mimeType '/json'}}
+						{{#if (isContentMultipart .)}}
+						if (req.files && !Array.isArray(req.files)) {
+							{{#with (toFileUploadProperties ../../requestBody.defaultContent.schema.properties)}}
```
We can lose the `#with` and just but the `(...)` into the `#each`.

karlvr replied:
And this should be:

```
{{#with (toFileUploadProperties schema.properties)}}
```

karlvr commented on src/index.ts:
```diff
@@ -55,6 +101,11 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
 
 			await emit('apiImpl', path.join(outputPath, relativeSourceOutputPath, 'impl', `${context.generator().toIdentifier(group.name)}.ts`), 
 				{ ...rootContext, ...group, ...doc }, false, hbs)
```
If we declare:

```typescript
const apiContainsMultipartOperations = containsMultipartOperations(operations)
```

Then we add property into this object, `{ containsMultipartOperations: apiContainsMultipartOperations }` then we don't need the `hasOperationSupportingMultipart` helper anymore... because in the template we can just check `{{#if containsMultipartOperations}}` (same where we generate `api.hbs`)

karlvr commented on src/index.ts:
```diff
@@ -40,6 +52,40 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
+					results.push({
+						name: prop,
+						minCount: null,
+						maxCount: null,
```
Should this be a `1` to validate that they don't submit two when we expect one?

Dakskihedron replied:
The 1 gets added in the `api.hbs` if `maxCount` is `null`, but it's probably a safer bet to do it here.

Dakskihedron committed:
chore: upgrade dependencies

Dakskihedron committed:
refactor: reduce helper usage

Introduce apiContainsMultipartOperations property and pass that to generator templates.

Dakskihedron committed:
feat: improve code robustness and naming

Rename toFileUploadProperties to fileUploadProperties and improve the type-checking in the helper. Also improve template where helper is used.

karlvr commented on src/index.ts:
```diff
@@ -40,21 +53,66 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
 			return !!value.mediaType.mimeType.match('\\bjson$')
 		})
 
+		hbs.registerHelper('hasOperationSupportingMultipart', function(value: CodegenOperation | CodegenOperation[]): boolean {
```
We can rename this to `operationSupportsMultipart` and it no longer needs to take an array parameter.

Dakskihedron committed:
feat: return 400 on unexpected field in multipart

karlvr commented on templates/apiMultipartHelper.hbs:
```diff
@@ -0,0 +1,6 @@
+/**
+ * File upload handling for multipart operations in the {{className name}} API endpoint.
+ */
+export const upload = multer({storage: multer.memoryStorage()})
```
Let's put the options onto their own line with a trailing comma... setting a good example.

```typescript
export const upload = multer({
    storage: multer.memoryStorage(),
})
```

karlvr commented on templates/validation.hbs:
```diff
@@ -303,5 +325,23 @@ export function timeToJson(name: string, value: string): string {
+	}
+}
+
+export function fileFromValue(name: string, value: Express.Multer.File[] | Express.Multer.File): any {
```
Does this ever get called with an array? If there is an array of files it should be an array in the schema, and we'd handle the array using `arrayFromJson`, which would ensure we're just called with a single file.

What happens if we remove the array from the type and just have this function just return `value`? Does everything still work?

Dakskihedron replied:
It gets called with an array only when the field supports a single file upload. For fields supporting file arrays, `arrayFromJson` is called on it so it ends up only taking file objects because the schema has the array type. Single file uploads use the object schema, so the array is retained.

When I removed the array type and return only `value`, the `request` ends up with the array containing the file object for single file uploads and remains normal for file arrays.

Dakskihedron replied:
For arrays of files, it gets called with the individual file objects because the top-level type of that schema is an array, so the generator knows to process it with `arrayFromJson` first.

karlvr replied:
@Dakskihedron thanks for this reasoning; it isn't clear from the implementation why this is, so we should add a comment in the `isArray` block below to explain it... OR that might trigger us to ask, why are we putting a file array into our request when we only want a single file and then having to deal with it here. That's daft.

SO

Let's make this function only accept a single file and ditch the array stuff.

Let's modify `api.hbs` so we pull out `[0]` from the `req.files["blah"]` if it's not an array property. You can add an `isArray` property into `FileUploadProperty` as a convenience.

Dakskihedron replied:
~~We can probably remove the `fileFromValue` function since its purpose was to extract the file object from the array.~~ We can't remove `fileFromValue` but we can rewrite it to type-check.

karlvr commented on templates/frag/extractValue.hbs:
```diff
@@ -0,0 +1,9 @@
+@param <this> CodegenProperty or CodegenSchema
+--}}
+{{#if (isFile schema)}}
+value
```
I need to re-understand why we needed to do this. It may be something we can fix up in `api.hbs` when we populate `req.body`... but I need to see the generated code to be sure.

Dakskihedron replied:
We do this because the validation function for file parts calls `fileFromValue`, so we ensure we pass in the array containing the file object or the individual file objects (if the field supports a file array upload).

Dakskihedron committed:
refactor: change naming and styling

Dakskihedron committed:
feat: set error response content-type

Dakskihedron committed:
refactor: undo separation of toJson functions

Dakskihedron committed:
refactor: remove extractValue frag dependency

karlvr commented on src/index.ts:
```diff
@@ -40,21 +53,66 @@ const createGenerator: CodegenGeneratorConstructor = (config, context) => {
 			return !!value.mediaType.mimeType.match('\\bjson$')
 		})
 
+		hbs.registerHelper('operationSupportingMultipart', function(value: CodegenOperation | CodegenOperation[]): boolean {
```
We haven't made the change suggested in my previous review that this no longer takes an array parameter.

Dakskihedron replied:
Whoops.

Dakskihedron committed:
refactor: remove unused logic from helper

Dakskihedron committed:
feat: improve type-checking for file objects
